src/DrDotnet.Profilers/profilers/gc_survivors_profiler.rs

```diff
@@ -28,251 +56,328 @@ impl Profiler for GCSurvivorsProfiler {
+                    // we care only for objects from GEN 2
+                    if gen.generation != ffi::COR_PRF_GC_GENERATION::COR_PRF_GC_GEN_2 {
+                        continue;
+                    }
+                }
+                Err(e) => { debug!("Error ({:?}) getting generation of object id: {object_id}", e); continue; }
+            }
+            for branch in self.append_references(info, object_id, 10) {
+                sequences.entry(branch)
+                    .and_modify(|referencers| {referencers.0.insert(object_id.clone(), size);})
+                    .or_insert(References(HashMap::from([(object_id.clone(), size)])));
             }
         }
 
-        info!("Successfully processed surviving roots :)");
+        info!("Graph built in {} ms", now.elapsed().as_millis());
 
-        // We're done, we can detach :)
-        let profiler_info = self.clr().clone();
-        profiler_info.request_profiler_detach(3000).ok();
-
-        Ok(())
+        sequences
     }
+    
+}
 
-    // https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling/icorprofilercallback2-rootreferences2-method
-    fn root_references_2(&mut self, root_ref_ids: &[ffi::ObjectID], root_kinds: &[ffi::COR_PRF_GC_ROOT_KIND], root_flags: &[ffi::COR_PRF_GC_ROOT_FLAGS], root_ids: &[ffi::UINT_PTR]) -> Result<(), ffi::HRESULT>
+impl CorProfilerCallback for GCSurvivorsProfiler
+{
+    fn object_references(&mut self, object_id: ffi::ObjectID, class_id: ffi::ClassID, object_ref_ids: &[ffi::ObjectID]) -> Result<(), HRESULT>
     {
-        // Only track in case of gen 2!
-
-        info!("Root references ({})", root_ids.len());
+        if !self.is_triggered_gc.load(Ordering::Relaxed) {
+            error!("Early return of object_references because GC wasn't forced yet");
+            // Early return if we received an event before the forced GC started
+            return Ok(());
+        }
 
-        for i in 0..root_ref_ids.len() {
-            let id = root_ref_ids[i];
-            //info!("Root '{}' of kind: {:?}", GCSurvivorsProfiler::get_object_class_name(self.profiler_info(), id), root_kinds[i]);
-            if id != 0 {
-                // For some reasons, this callback may return several times the same object references, so we use a duplication free collection
-                self.root_references.insert(id);
-            }
+        // Create dependency tree, but from object to referencers, instead of object to its references.
+        // This is usefull for being able to browse from any object back to its roots.
+        for object_ref_id in object_ref_ids {
+            self.object_to_referencers.entry(*object_ref_id)
+                .and_modify(|referencers| referencers.push(object_id))
+                .or_insert(vec![object_id]);
         }
 
+        // Also add this object, with no referencers, just in case this object isn't referenced 
+        self.object_to_referencers.insert(object_id, vec![]);
```

Comment:
Nice catch!! Was it the reason why whe didn't saw any SurvivorObject in our tests? Because they were never referenced?
Another bug was in case `object_id` references nothing (`object_ref_ids `is empty) we previously would have skipped it.